File: lintcode_algo_2021/s7.-ds-design/130.Heapify/130_heapify__shiftdown.java
public class Solution {
    /*
     * @param A: Given an integer array
     * @return: nothing
     */
    public void heapify(int[] A) {
        // write your code here
        for (int i = (A.length - 2)/2; i >= 0; i--) {
            shiftDown(A, i);
        }
    }

    private void shiftDown(int[] A, int index) {
        int n = A.length;
        while (index < n) {
            int right = index * 2 + 1;
            int left = index * 2 + 2;
            int minIndex = index;

            if (left < n && A[left] < A[minIndex]) {
                minIndex = left;
            }

            if (right < n && A[right] < A[minIndex]) {
                minIndex = right;
            }

            if (index == minIndex) {
                break;
            }

            int tmp = A[minIndex];
            A[minIndex] = A[index];
            A[index] = tmp;

            index = minIndex;
        }
    }
}

File: lintcode_algo_2021/s7.-ds-design/130.Heapify/130_heapify__shiftdown.py
class Solution:
    """
    @param: A: Given an integer array
    @return: nothing
    """
    def heapify(self, A):
        for i in range(len(A)//2, -1, -1):
            self.down_shift(A, i)
    
    def down_shift(self, A, i):
        n = len(A)
        while i < n:
            left_id = i * 2 + 1
            right_id = i * 2 + 2 
            min_id = i
            if left_id < n and A[left_id] < A[min_id]:
                min_id = left_id
            if right_id < n and A[right_id] < A[min_id]:
                min_id = right_id
            
            if min_id == i:
                break

            A[i], A[min_id] = A[min_id], A[i]
            
            i = min_id